extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +150,76 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticSnapshotsMap = new Map<vscode.Uri, readonly vscode.Diagnostic[]>();
+	private _timeout: NodeJS.Timeout | undefined;
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _getDiagnostics: (uri: vscode.Uri) => readonly vscode.Diagnostic[]
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript') {
+				this._updateDiagnosticCodesAfterTimeoutForUri(e.document.uri);
+			}
+		}));
+		this._updateAllDiagnosticCodesAfterTimeout();
+		this._registerTelemetryEventEmitter();
+	}
+
+	private _updateAllDiagnosticCodesAfterTimeout() {
+		const uris = vscode.workspace.textDocuments.map(doc => doc.uri);
+		uris.forEach(uri => setTimeout(() => { this._updateDiagnosticCodes(uri); }, 10000));
+	}
+
+	private _updateDiagnosticCodesAfterTimeoutForUri(uri: vscode.Uri) {
+		clearTimeout(this._timeout);
+		this._timeout = setTimeout(() => { this._updateDiagnosticCodes(uri); }, 10000);
```

Comment:
This pattern has a bug where if a timer is installed for uri1 and then the user types in uri2, then the timer for uri1 is canceled and uri1 is not updated anymore. I suggest to update all diagnostics in all documents when the timer fires.

Also, 10s seems like a large number to me.